hal/hal/emac_api.h

```diff
@@ -0,0 +1,173 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_EMAC_API_H
+#define MBED_EMAC_API_H
+
+#include "platform.h"
+
+#if DEVICE_EMAC
+
+#include "emac_stack_mem.h"
+
+typedef struct emac_interface emac_interface_t;
+
+/**
+ * EmacInterface
+ *
+ * This interface should be used to abstract low level access to networking hardware
+ */
+
+/**
+ * Network interface capabilities
+ *
+ * Right now the flags are LWIP specific, but in the future we may need to support more stacks.
+ */
+#define EMAC_CAP_BROADCAST    0x01U     /*!< Interface has broadcast capabilities */
+#define EMAC_CAP_ETHARP       0x02U     /*!< Interface supports ARP */
+#define EMAC_CAP_ETHERNET     0x04U     /*!< Ethernet device */
+#define EMAC_CAP_IGMP         0x08U     /*!< Interface has IGMP capabilities */
+
+/**
+ * Callback to be register with Emac interface and to be called fore received packets
+ *
+ * @param data Arbitrary user data (IP stack)
+ * @param buf  Received data
+ */
+typedef void (*emac_link_input_fn)(void *data, emac_stack_mem_chain_t *buf);
+
+/**
+ * Callback to be register with Emac interface and to be called for link status changes
+ *
+ * @param  data Arbitrary user data (IP stack)
+ * @param  up   Link status
+ */
+typedef void (*emac_link_state_change_fn)(void *data, bool up);
+
+/**
+ * Return maximum transmission unit
+ *
+ * @param emac Emac interface
+ * @return     MTU in bytes
+ */
+typedef uint32_t (*emac_get_mtu_size_fn)(emac_interface_t *emac);
+
+/**
+ * Return default host name for the interface
+ *
+ * @param emac Emac interface
+ * @return     Default host name
+ */
+typedef const char* (*emac_get_hostname_fn)(emac_interface_t *emac);
+
```

Comment:
Having the driver set the host name never made any sense, IMO. It should be set by the application, if specified at all. I guess it would make sense if there was a private application->driver hook behind this, but that feels weird.

AFAICT, the host name isn't currently used for anything in lwIP, except it can optionally pass it to the DHCPv4 server in an option.